design-documents/webhooks.md

```diff
@@ -0,0 +1,160 @@
+## Webhooks
+
+### Terms
+
+A webhook (also called a web callback or HTTP push API) is a way for an app to provide other applications with real-time information.
+
+### Overview
+As Magento is becoming more and more a composition of internal APIs and external services, Rest and GraphQL APIs are not able cover every real-life use case. Traditional Web APIs allow to retrieve or post content in a programmatical way, far from being optimal when you just want your application to be told when something changed. If that is the case, it would require polling of whole data at regular intervals, and that just doesn’t scale.
+
+![pull all data approach](img/traditional_pull_all_approach.png)
+
+Polling a whole API's data is generally a wasteful and messy way of trying to retrieve updates.
+
+Traditionally, Magento developers switch to push approach at this point. This approach supposes a separate integration specific extension, that is installed on Magento instance and pushes changed data into specific service.
+
+![multi_extensions_push](img/multi_extension_push.png)
+
+This approach definitely improves communication scheme by sending only changed data. However, the scheme itself became much more complex and it needs to be implemented in every extension. Also, we're still wasting resources on multiple queues and multiple "extension's data collection layers". Last point is mostly visible on store-front data, where you need to pass data through business logic in order to get the final result. 
+
+On this scheme each extension is doing the same job for the same data, but in most cases it will be done in different ways producing some mess and wasting server resources.
+
+Of course this scheme also introduces a huge field for mistakes in authorization and authentication layers, because they need to be implemented in every extension.
+
+The similar issues affect in-process communications as well as network communcations. A good example is full text search, which needs up-to-date store-front data and hooks in all possible places in order to track the changes.
+
+Webhooks are aimed to solve exactly this class of issues and the underlying notification layer will solve in-process notification issues 
+
+### Design
+
+The principle scheme of catalog `webhooks` in context of services isolation environment:
+
+![webhooks net flow](img/webhooks_flow.png)
+
+Simpler example:
+* Customer register on storefront (Command API)
+* Command API emits `after_transaction_commit` event providing customer id after committing customer's data
+* Webhook listener calls customer query API with provided customer id. The response may contain new data that was not provided in original request(e.g. totalSpent, customer_group, total_orders, etc)
+* Webhook framework notifies all subscribers with customer data
+
+:warning: Webhooks are supposed to push data updates only. Initial data should be extracted from Magento using standard APIs like GrapQL and REST. :warning:
```

Comment:
I've done quite a bit of freelance work and this would render this feature nearly completely useless for the ways that most clients would want to use it. For example, having a Zap that adds a customer to a CRM workflow when they are created. Or having a customer unsubscribed from a mailing list when they are deleted.

Reply:
This point is not about partial data, but more about the initial system state(e.g. send all existing products using webhooks at the moment of subscription). I'll remove this point in order to avoid any confusion.